x-pack/legacy/plugins/siem/public/pages/detection_engine/components/user_info/index.tsx

```diff
@@ -0,0 +1,185 @@
+
+type Return = [boolean, boolean | null, boolean | null, boolean | null, string | null];
+
+export interface State {
+  canUserCRUD: boolean | null;
+  hasWrite: boolean | null;
+  isSignalIndexExists: boolean | null;
+  isAuthenticated: boolean | null;
+  loading: boolean | null;
+  signalIndexName: string | null;
+}
+
+const initialState: State = {
+  canUserCRUD: null,
+  hasWrite: null,
+  isSignalIndexExists: null,
+  isAuthenticated: null,
+  loading: true,
+  signalIndexName: null,
+};
+
+export type Action =
+  | { type: 'updateLoading'; loading: boolean }
+  | {
+      type: 'updateHasWrite';
+      hasWrite: boolean | null;
+    }
+  | {
+      type: 'updateIsSignalIndexExists';
+      isSignalIndexExists: boolean | null;
+    }
+  | {
+      type: 'updateIsAuthenticated';
+      isAuthenticated: boolean | null;
+    }
+  | {
+      type: 'updateCanUserCRUD';
+      canUserCRUD: boolean | null;
+    }
+  | {
+      type: 'updateSignalIndexName';
+      signalIndexName: string | null;
+    };
+
+export const userInfoReducer = (state: State, action: Action): State => {
+  switch (action.type) {
+    case 'updateLoading': {
+      return {
+        ...state,
+        loading: action.loading,
+      };
+    }
+    case 'updateHasWrite': {
+      return {
+        ...state,
+        hasWrite: action.hasWrite,
+      };
+    }
+    case 'updateIsSignalIndexExists': {
+      return {
+        ...state,
+        isSignalIndexExists: action.isSignalIndexExists,
+      };
+    }
+    case 'updateIsAuthenticated': {
+      return {
+        ...state,
+        isAuthenticated: action.isAuthenticated,
+      };
+    }
+    case 'updateCanUserCRUD': {
+      return {
+        ...state,
+        canUserCRUD: action.canUserCRUD,
+      };
+    }
+    case 'updateSignalIndexName': {
+      return {
+        ...state,
+        signalIndexName: action.signalIndexName,
+      };
+    }
+    default:
+      return state;
+  }
+};
+
+const StateUserInfoContext = createContext<[State, Dispatch<Action>]>([initialState, () => noop]);
+
+const useUserData = () => useContext(StateUserInfoContext);
+
+interface ManageUserInfoProps {
+  children: React.ReactNode;
+}
+
+export const ManageUserInfo = ({ children }: ManageUserInfoProps) => (
+  <StateUserInfoContext.Provider value={useReducer(userInfoReducer, initialState)}>
+    {children}
+  </StateUserInfoContext.Provider>
+);
+
+export const useUserInfo = (): Return => {
+  const [
+    { canUserCRUD, hasWrite, isSignalIndexExists, isAuthenticated, loading, signalIndexName },
+    dispatch,
+  ] = useUserData();
+  const [privilegeLoading, isApiAuthenticated, hasApiWrite] = usePrivilegeUser();
+  const [
+    indexNameLoading,
+    isApiSignalIndexExists,
+    apiSignalIndexName,
+    createSignalIndex,
+  ] = useSignalIndex();
+
+  const uiCapabilities = useKibana().services.application?.capabilities;
+  const capabilitiesCanUserCRUD = (uiCapabilities?.siem?.crud as boolean) ?? false;
```

Comment:
Could of Q's here:

For:
```ts
const uiCapabilities = useKibana().services.application?.capabilities;
```

Are the types incorrect? Because I was able to just do this:

```ts
const uiCapabilities = useKibana().services.application.capabilities;
```

according to the types.

For this line:
```
const capabilitiesCanUserCRUD = (uiCapabilities?.siem?.crud as boolean) ?? false;
```

Are parts of those types wrong? uiCapabilities above doesn't say it can be null or undefined but right below we begin just checking for it using the existential operator. The `as cast` and assuming it is a boolean is a bit 

I was able to use just this code and seemed to be ok?

```ts
  const uiCapabilities = useKibana().services.application.capabilities;
  const capabilitiesCanUserCRUD: boolean =
    typeof uiCapabilities.siem.crud === 'boolean' ? uiCapabilities.siem.crud : false;
```

I might be missing more details but am just cautious about the compile time type types drifting too much from the runtime types.